test/testother.cpp

```diff
@@ -2051,6 +2051,187 @@ class TestOther : public TestFixture {
+              "    x.dostuff();\n"
+              "    U& y = dynamic_cast<U&>(x)\n"
+              "    y.mutate();\n" //to avoid warnings that y can be const
+              "}");
+        ASSERT_EQUALS("", errout.str());
+        check("struct T : public U { void dostuff() const {}};\n"
+            "void a(T& x) {\n"
+            "    x.dostuff();\n"
+            "    const U& y = dynamic_cast<typename const U&>(x)\n"
+            "}");
+        ASSERT_EQUALS("[test.cpp:2]: (style) Parameter 'x' can be declared with const\n", errout.str());
+        check("struct T : public U { void dostuff() const {}};\n"
+            "void a(T& x) {\n"
+            "    x.dostuff();\n"
+            "    U& y = dynamic_cast<typename U&>(x)\n"
+            "    y.mutate();\n" //to avoid warnings that y can be const
+            "}");
+        ASSERT_EQUALS("", errout.str());
+        check("struct T : public U { void dostuff() const {}};\n"
+            "void a(T& x) {\n"
+            "    x.dostuff();\n"
+            "    U* y = dynamic_cast<U*>(&x)\n"
+            "    y->mutate();\n" //to avoid warnings that y can be const
+            "}");
+        ASSERT_EQUALS("", errout.str());
+
+        check("struct T : public U { void dostuff() const {}};\n"
+            "void a(T& x) {\n"
+            "    x.dostuff();\n"
+            "    const U * y = dynamic_cast<const U *>(&x)\n"
+            "    y->mutate();\n" //to avoid warnings that y can be const
+            "}");
+        TODO_ASSERT_EQUALS("can be const", errout.str(), ""); //Currently taking the address is treated as a non-const operation when it should depend on what we do with it
+        check("struct T : public U { void dostuff() const {}};\n"
+            "void a(T& x) {\n"
+            "    x.dostuff();\n"
+            "    U const * y = dynamic_cast<U const *>(&x)\n"
+            "    y->mutate();\n" //to avoid warnings that y can be const
+            "}");
+        TODO_ASSERT_EQUALS("can be const", errout.str(), ""); //Currently taking the address is treated as a non-const operation when it should depend on what we do with it
+        check("struct T : public U { void dostuff() const {}};\n"
+            "void a(T& x) {\n"
+            "    x.dostuff();\n"
+            "    U * const y = dynamic_cast<U * const>(&x)\n"
+            "    y->mutate();\n" //to avoid warnings that y can be const
+            "}");
+        ASSERT_EQUALS("", errout.str());
+        check("struct T : public U { void dostuff() const {}};\n"
+            "void a(T& x) {\n"
+            "    x.dostuff();\n"
+            "    const U const * const * const * const y = dynamic_cast<const U const * const * const * const>(&x)\n"
+            "    y->mutate();\n" //to avoid warnings that y can be const
+            "}");
+        TODO_ASSERT_EQUALS("can be const", errout.str(), ""); //Currently taking the address is treated as a non-const operation when it should depend on what we do with it
```

Comment:
My code no longer cares whether you're casting to a pointer or reference but it won't actually be executed in this case because taking the address is enough to treat it as non-const.

Reply:
And if the original variable is a pointer then it will fail the isReferenceCheck at https://github.com/danmar/cppcheck/blob/8e79b0c8bc825d94446432f4d00c761e395b85c7/lib/checkother.cpp#L1341